pdns/syncres.cc

```diff
@@ -637,10 +637,57 @@ int SyncRes::asyncresolveWrapper(const ComboAddress& ip, bool ednsMANDATORY, con
   return ret;
 }
 
+bool SyncRes::qnameRPZHit(const DNSFilterEngine& dfe, DNSName& target, const QType& qtype, vector<DNSRecord> &ret)
+{
+  if (!d_wantsRPZ) {
+    return false;
+  }
+
+  bool match = dfe.getQueryPolicy(target, d_discardedPolicies, d_appliedPolicy, true);
+  if (!match) {
+    return false;
+  }
+
+  mergePolicyTags(d_policyTags, d_appliedPolicy.getTags());
+  if (d_appliedPolicy.d_kind == DNSFilterEngine::PolicyKind::NoAction) {
+    return false;
+  }
+  LOG(": (hit by RPZ policy '" + d_appliedPolicy.getName() + "')" << endl);
+  if (d_appliedPolicy.d_kind == DNSFilterEngine::PolicyKind::Truncate) {
+    // XXX We don't know if we're doing TCP here....
+    return false;
+  }
+  if (d_appliedPolicy.d_kind != DNSFilterEngine::PolicyKind::Custom) {
+    return true;
+  }
+  auto spoofed = d_appliedPolicy.getCustomRecords(target, qtype.getCode());
+  for (const auto& dr : spoofed) {
+    if (dr.d_place != DNSResourceRecord::ANSWER) {
+      continue;
+    }
+    ret.push_back(dr);
+    auto content = getRR<CNAMERecordContent>(dr);
+    if (content) {
+      target = content->getTarget();
+      return qnameRPZHit(dfe, target, qtype, ret);
```

Comment:
So this would chase CNAME targets but only if they exist in a policy, this would not actually chase all CNAMEs, right? I understand this is just a start, I want to make sure I understand whether you intend to chase all CNAMEs in the end :)

Reply:
I nthin k it is ok to keep chasing withing the RPZ, the resulting new target will be further resolved by the caller. I just need to make sure we return `false` in that case. So that is an actual bug.